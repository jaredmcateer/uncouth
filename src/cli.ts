import program from "commander";
import { convertFile } from "./index.js";
import inquirer from "inquirer";
import { writeFileInfo } from "./file";
import { version } from "../package.json";

function camelize(str: string) {
  return str.replace(/-(\w)/g, (_, c: string) => (c ? c.toUpperCase() : ""));
}

function getCmdOptions(cmd: { options: Array<{ long: string }> }) {
  const args: { [key: string]: boolean | string } = {};
  cmd.options.forEach((o: { long: string }) => {
    const key = camelize(o.long.replace(/^--/, ""));

    if (
      typeof (cmd as unknown as Record<string, string>)[key] !== "function" &&
      typeof (cmd as unknown as Record<string, string>)[key] !== "undefined"
    ) {
      args[key] = (cmd as unknown as Record<string, string>)[key];
    }
  });
  return args;
}

program.version(version).usage("<command> [options]");

program
  .command("single <filePath>")
  .description("convert vue component file from class to composition api")
  .option("-v, --view", "Output file content on stdout, and no write file.")
  .option("-o, --output", "Output result file path.")
  .option("-r, --root <root>", "Set root path for calc file absolute path. Default:`process.cwd()`")
  .option("-c, --config <config>", "Set uncouth config file path. Default: `'.uncouth.js'`")
  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  .action(async (filePath: string, cmd) => {
    const cmdOptions = getCmdOptions(cmd);
    if (!cmdOptions.output && !cmdOptions.view) {
      const result = await inquirer.prompt({
        name: "ok",
        type: "confirm",
        message:
          "You aren't using -o option to set output file path, It will replace original file content.",
      });
      if (!result.ok) {
        return;
      }
    }

    const { file, result } = convertFile(
      filePath,
      cmdOptions.root as string,
      cmdOptions.config as string
    );
    if (cmdOptions.view) {
      console.log(result);
      return;
    }

    writeFileInfo(file, result);
    console.log("Please check the TODO comments on result.");
  });

program.parse(process.argv);
