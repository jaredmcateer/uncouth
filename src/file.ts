import path from "path";
import fs from "fs";
import { isVueFile, parseVueFile } from "./utils";
import { UncouthOptions } from "./options";
import { log } from "./debug";

export enum FileKind {
  VUE,
  TS,
}

export type FileInfo =
  | {
      fsPath: string;
      kind: FileKind;
      content: string;
    }
  | {
      fsPath: string;
      kind: FileKind;
      fileContent: string;
      start: number;
      end: number;
      content: string;
    };

export function readVueSFCOrTsFile(filePath: string, options: UncouthOptions): FileInfo {
  const fileFsPath = path.isAbsolute(filePath) ? filePath : path.resolve(options.root, filePath);
  const fileContent = fs.readFileSync(fileFsPath, { encoding: "utf8" });
  if (isVueFile(fileFsPath)) {
    const scriptContent = parseVueFile(options.vueTemplateCompiler, fileContent).script;
    if (scriptContent) {
      log(`Readed Vue file: ${fileFsPath}`);
      return {
        fsPath: fileFsPath,
        kind: FileKind.VUE,
        start: scriptContent.start,
        end: scriptContent.end,
        content: scriptContent.content,
        fileContent,
      };
    }
    throw new Error("The Vue SFC don't have sciprt element.");
  } else {
    log(`Readed TS file: ${fileFsPath}`);
    return {
      fsPath: fileFsPath,
      kind: FileKind.TS,
      content: fileContent,
    };
  }
}

export function writeFileInfo(fileInfo: FileInfo, content: string): void {
  if ("start" in fileInfo) {
    log(`Write Vue file: ${fileInfo.fsPath}`);
    const fileContent = `${fileInfo.fileContent.slice(
      0,
      fileInfo.start
    )}\n${content}${fileInfo.fileContent.slice(fileInfo.end)}`;
    fs.writeFileSync(fileInfo.fsPath, fileContent, { encoding: "utf8" });
  } else {
    log(`Write TS file: ${fileInfo.fsPath}`);
    fs.writeFileSync(fileInfo.fsPath, content, { encoding: "utf8" });
  }
}

export function existsFileSync(path: string): boolean {
  return fs.existsSync(path);
}
