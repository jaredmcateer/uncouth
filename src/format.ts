/* eslint-disable @typescript-eslint/no-var-requires */
import { UncouthOptions } from "./options";
import path from "path";
import { log } from "./debug";
import prettier from "prettier/standalone";
import prettierTypescriptParser from "prettier/parser-typescript";
import { existsFileSync } from "./file";

export function format(content: string, options: UncouthOptions): string {
  const isNode = typeof window === "undefined";
  if (!isNode) {
    return prettier.format(content, {
      plugins: [prettierTypescriptParser],
      parser: "typescript",
      semi: false,
      singleQuote: true,
    });
  }

  const eslintConfigPath = path.resolve(options.root, options.eslintConfigFile);
  const prettierFormat = require("prettier-eslint") as (config: unknown) => string;
  const prettierEslintOpions = existsFileSync(eslintConfigPath)
    ? {
        text: content,
        filePath: eslintConfigPath,
        prettierOptions: {
          parser: "typescript",
        },
        fallbackPrettierOptions: {
          parser: "typescript",
        },
      }
    : {
        text: content,
        filePath: "",
        eslintConfig: {
          parser: require.resolve("@typescript-eslint/parser"),
          parserOptions: {
            sourceType: "module",
            ecmaFeatures: {
              jsx: false,
            },
          },
          rules: {
            semi: ["error", "never"],
            "padding-line-between-statements": [
              "error",
              { blankLine: "always", prev: "*", next: "export" },
              { blankLine: "always", prev: "const", next: "*" },
              { blankLine: "always", prev: "*", next: "const" },
            ],
          },
        },
        prettierOptions: {
          parser: "typescript",
          Semicolons: false,
          singleQuote: true,
          trailingComma: "none",
        },
        fallbackPrettierOptions: {
          parser: "typescript",
          singleQuote: true,
          Semicolons: false,
          trailingComma: "none",
        },
      };

  log("Format result code.....");
  return prettierFormat(prettierEslintOpions);
}
