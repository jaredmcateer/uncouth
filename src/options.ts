import ts from "typescript";
import { ASTConvertPlugins } from "./plugins/types";
import { getDefaultPlugins } from "./plugins";
import * as vueTemplateCompiler from "vue-template-compiler";

export interface UncouthOptions {
  root: string;
  debug: boolean;
  compatible: boolean;
  setupPropsKey: string;
  setupContextKey: string;
  typescript: typeof ts;
  vueTemplateCompiler: typeof vueTemplateCompiler;
  prettierConfig: string;
  plugins: ASTConvertPlugins;
}

export type InputUncouthOptions = Partial<UncouthOptions>;

export function getDefaultUncouthOptions(tsModule: typeof ts = ts): UncouthOptions {
  return {
    root: process.cwd(),
    debug: false,
    compatible: false,
    setupPropsKey: "props",
    setupContextKey: "context",
    typescript: tsModule,
    vueTemplateCompiler: vueTemplateCompiler,
    prettierConfig: ".prettierrc",
    plugins: getDefaultPlugins(tsModule),
  };
}

export function mergeUncouthOptions(
  original: UncouthOptions,
  merged: InputUncouthOptions
): UncouthOptions {
  return {
    ...original,
    ...merged,
  };
}
