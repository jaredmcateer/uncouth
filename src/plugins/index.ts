import ts from "typescript";
import { UncouthOptions } from "../options";
import { ASTConvertPlugins, ASTResult, ASTConverter, ASTResultKind } from "./types";
import { convertNodeToASTResult } from "../utils";
import { log } from "../debug";
import { convertObjName } from "./vue-class-component/object/ComponentName";
import { convertObjProps } from "./vue-class-component/object/Prop";
import { convertObjData } from "./vue-class-component/object/Data";
import { convertName, mergeName } from "./vue-class-component/ComponentName";
import { convertModel } from "./vue-property-decorator/Model";
import { convertProp, mergeProps } from "./vue-property-decorator/Prop";
import { convertDomRef } from "./vue-property-decorator/DomRef";
import { convertData } from "./vue-class-component/Data";
import { convertGetter, convertSetter, mergeComputed } from "./vue-class-component/Computed";
import { convertIntervalHook } from "./vue-class-component/IntervalHook";
import { convertWatch } from "./vue-property-decorator/Watch";
import { convertEmitMethod } from "./vue-property-decorator/Emit";
import { convertMethod } from "./vue-class-component/Method";
import { changeContextAndSort } from "./changeContextAndSort";
import { convertRender } from "./vue-class-component/Render";
import { convertInject } from "./vue-property-decorator/Inject";
import { convertProvide } from "./vue-property-decorator/Provide";
import { convertTemplateRef } from "./vue-class-component/TemplateRef";
import { TsHelper } from "../helpers/TsHelper";
import { convertVuexAction } from "./vuex-class/Action";
import { convertVuexGetter } from "./vuex-class/Getter";

export function getDefaultPlugins(tsModule: typeof ts): ASTConvertPlugins {
  return {
    [tsModule.SyntaxKind.Decorator]: {
      [tsModule.SyntaxKind.PropertyAssignment]: [convertObjName, convertObjProps],
      [tsModule.SyntaxKind.MethodDeclaration]: [convertObjData, convertIntervalHook],
    },
    [tsModule.SyntaxKind.Identifier]: [convertName],
    [tsModule.SyntaxKind.HeritageClause]: [
      // TODO: extends Minix
    ],
    [tsModule.SyntaxKind.PropertyDeclaration]: [
      convertModel,
      convertProp,
      convertDomRef,
      convertProvide,
      convertInject,
      convertData,
      convertTemplateRef,
      convertVuexAction,
      convertVuexGetter,
    ],
    [tsModule.SyntaxKind.GetAccessor]: [convertGetter],
    [tsModule.SyntaxKind.SetAccessor]: [convertSetter],
    [tsModule.SyntaxKind.MethodDeclaration]: [
      convertRender,
      convertIntervalHook,
      convertWatch,
      convertEmitMethod,
      convertMethod,
    ],
    after: [mergeName, mergeProps, mergeComputed, changeContextAndSort],
  };
}

export function getDecoratorArgumentExpr(
  tsModule: typeof ts,
  node: ts.Node
): ts.ObjectLiteralExpression | undefined {
  if (tsModule.isCallExpression(node)) {
    if (node.arguments.length > 0) {
      return node.arguments[0] as ts.ObjectLiteralExpression;
    }
  }

  return undefined;
}

export function getASTResults(
  node: ts.ClassDeclaration,
  options: UncouthOptions,
  program: ts.Program
): ASTResult<ts.Node>[] {
  const $t = new TsHelper(options);
  const converterPlugins = options.plugins;

  let astResults: ASTResult<ts.Node>[] = [];
  node.forEachChild((child) => {
    if ($t.module.isDecorator(child)) {
      const objExpr = getDecoratorArgumentExpr($t.module, child.expression);
      if (objExpr) {
        objExpr.forEachChild((property) => {
          if (property.kind in converterPlugins[$t.module.SyntaxKind.Decorator]) {
            const objConverters = (
              converterPlugins[$t.module.SyntaxKind.Decorator] as unknown as {
                [index: number]: Array<ASTConverter<ts.Node>>;
              }
            )[property.kind];
            let converted = false;
            for (const converter of objConverters) {
              const result = converter(property, options, program);
              if (result) {
                astResults.push(result);
                converted = true;
                break;
              }
            }
            if (!converted) {
              astResults.push(convertNodeToASTResult($t, property));
            }
          }
        });
      }
    } else {
      if (child.kind in converterPlugins) {
        const converters = (
          converterPlugins as unknown as { [index: number]: Array<ASTConverter<ts.Node>> }
        )[child.kind];
        for (const converter of converters) {
          const result = converter(child, options, program);
          if (result) {
            astResults.push(result);
            break;
          }
        }
      }
    }
  });

  for (const fn of converterPlugins.after) {
    astResults = fn(astResults, options, program);
  }

  return astResults;
}

export function convertASTResultToSetupFn(
  astResults: ASTResult<ts.Node>[],
  options: UncouthOptions
): ts.MethodDeclaration {
  const $t = new TsHelper(options);

  const composables = getComposables(astResults, $t);

  const returnStatement = $t.addTodoComment(
    $t.factory.createReturnStatement(
      $t.factory.createObjectLiteralExpression([
        ...astResults
          .filter((el) => el.kind === ASTResultKind.COMPOSITION)
          .reduce((array, el) => array.concat(el.attributes), [] as string[])
          .map((el) => $t.factory.createShorthandPropertyAssignment(el, undefined)),
      ])
    ),
    "Please remove unused return variable",
    false
  );

  return $t.createMethod(
    "setup",
    [options.setupPropsKey, options.setupContextKey],
    [
      ...composables,
      ...(astResults
        .filter((el) => el.kind === ASTResultKind.COMPOSITION)
        .map((el) => el.nodes)
        .reduce((array, el) => array.concat(el), []) as ts.Statement[]),
      returnStatement,
    ]
  );
}

interface Clause {
  named: Set<string>;
  default?: string;
  params?: ts.Expression[];
}

function getComposables(astResults: ASTResult<ts.Node>[], $t: TsHelper): ts.VariableStatement[] {
  const composableMap = new Map<string, Clause>();
  astResults.forEach((result) => {
    if (!result.composables) return;

    result.composables.forEach((info) => {
      const func = info.func;
      const tmp: Clause = composableMap.get(func) ?? { named: new Set() };

      if (!("default" in tmp) && "default" in info) {
        tmp.default = info.default;
      }

      if ("params" in info) {
        tmp.params = info.params;
      }

      if ("named" in info) {
        info.named?.forEach((name) => tmp.named.add(name));
      }

      composableMap.set(func, tmp);
    });
  });

  const composables = Array.from(composableMap)
    .map(([func, clause]) => {
      const funcExpr = $t.createCallExpression(func, undefined, clause.params);
      const statements: ts.VariableStatement[] = [];
      if (clause.default) {
        statements.push($t.createConstStatement(clause.default, funcExpr));
      } else {
        const u = undefined;
        const importElements = [...clause.named].map((name) =>
          $t.factory.createBindingElement(u, u, $t.factory.createIdentifier(name))
        );

        if (importElements.length > 0) {
          const vars = $t.factory.createObjectBindingPattern(importElements);
          const boundElementsDeclaration = $t.factory.createVariableDeclaration(
            vars,
            u,
            u,
            funcExpr
          );
          const varDecList = $t.factory.createVariableDeclarationList([boundElementsDeclaration]);
          statements.push($t.factory.createVariableStatement(u, varDecList));
        }
      }

      return statements;
    })
    .flat()
    .filter((composable): composable is ts.VariableStatement => !!composable);

  return composables;
}

export function convertASTResultToImport(
  astResults: ASTResult<ts.Node>[],
  options: UncouthOptions
): ts.ImportDeclaration[] {
  const $t = new TsHelper(options);

  const importMap = new Map<string, Clause>();
  for (const result of astResults) {
    for (const importInfo of result.imports) {
      const key: string = "external" in importInfo ? importInfo.external : importInfo.path;
      const temp: Clause = importMap.get(key) ?? { named: new Set() };
      if (!("default" in temp) && "default" in importInfo) {
        temp.default = importInfo.default;
      }

      if (key === "vue") {
        temp.named.add("defineComponent");
      }

      for (const named of importInfo.named || []) {
        temp.named.add(named);
      }

      importMap.set(key, temp);
    }
  }

  if (options.compatible && importMap.has("@vue/composition-api")) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const temp = importMap.get("@vue/composition-api")!;
    temp.named.add("defineComponent");
    importMap.set("@vue/composition-api", temp);
  }

  return Array.from(importMap).map((el) => {
    const [key, clause] = el;

    const name = clause.default ? $t.factory.createIdentifier(clause.default) : undefined;
    const importElements = [...clause.named].map((named) =>
      $t.factory.createImportSpecifier(false, undefined, $t.factory.createIdentifier(named))
    );
    const namedImports = $t.factory.createNamedImports(importElements);

    return $t.factory.createImportDeclaration(
      undefined,
      undefined,
      $t.factory.createImportClause(false, name, namedImports),
      $t.factory.createStringLiteral(key)
    );
  });
}

export function runPlugins(
  node: ts.ClassDeclaration,
  options: UncouthOptions,
  program: ts.Program
): ts.Statement[] {
  const $t = new TsHelper(options);
  log("Start Run ASTPlugins");
  const results = getASTResults(node, options, program);
  log("Finished ASTPlugins");

  log("Make setup function");
  const setupFn = convertASTResultToSetupFn(results, options);
  log("Make default export object");
  const dcIdentifier = $t.factory.createIdentifier("defineComponent");
  const dcOptionsProperties = [
    ...(results
      .filter((el) => el.kind === ASTResultKind.OBJECT)
      .map((el) => el.nodes)
      .reduce((array, el) => array.concat(el), []) as ts.PropertyAssignment[]),
    setupFn,
  ];
  const dcOptions = $t.factory.createObjectLiteralExpression(dcOptionsProperties, true);
  const dcExpression = $t.createCallExpression(dcIdentifier, undefined, [dcOptions]);
  const dcExportAssignment = $t.factory.createExportAssignment(
    undefined,
    undefined,
    undefined,
    dcExpression
  );

  const defineComponentExportAssignment = $t.copySyntheticComments(dcExportAssignment, node);

  log("Make ImportDeclaration");
  const importDeclaration = convertASTResultToImport(results, options);

  return [...importDeclaration, defineComponentExportAssignment];
}
