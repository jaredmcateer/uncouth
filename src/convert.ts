import type ts from "typescript";
import { getDecoratorNames, getDefaultExportNode } from "./utils";
import { runPlugins } from "./plugins";
import { UncouthOptions } from "./options";
import { log } from "./debug";

const vueClassModules = ["vue-class-component", "vue-property-decorator"];

export function convertAST(
  sourceFile: ts.SourceFile,
  options: UncouthOptions,
  program: ts.Program
): string {
  const tsModule = options.typescript;

  log("check vue class library");
  const vueClassModuleImportStatement = sourceFile.statements.find((statement) => {
    if (tsModule.isImportDeclaration(statement)) {
      if (vueClassModules.includes((statement.moduleSpecifier as ts.StringLiteral).text)) {
        return true;
      }
    }
    return false;
  });
  if (!vueClassModuleImportStatement) {
    throw new Error("no vue class library in this file.");
  }

  log("check default export class");
  const defaultExportNode = getDefaultExportNode(options.typescript, sourceFile);
  if (!defaultExportNode) {
    throw new Error("no default export class");
  }

  const otherStatements = sourceFile.statements
    .map((el) => el)
    .filter(
      (el) =>
        !(
          (tsModule.isClassDeclaration(el) &&
            getDecoratorNames(tsModule, el).includes("Component")) ||
          (tsModule.isImportDeclaration(el) &&
            vueClassModules.includes((el.moduleSpecifier as ts.StringLiteral).text)) ||
          (tsModule.isImportDeclaration(el) &&
            (el.moduleSpecifier as ts.StringLiteral).text === "vue")
        )
    );

  let resultStatements = [...otherStatements, ...runPlugins(defaultExportNode, options, program)];

  resultStatements = [
    ...resultStatements.filter((el) => tsModule.isImportDeclaration(el)),
    ...resultStatements.filter((el) => !tsModule.isImportDeclaration(el)),
  ];

  log("output result code");
  const printer = tsModule.createPrinter();
  const { factory } = tsModule;
  const result = printer.printFile(factory.updateSourceFile(sourceFile, resultStatements));

  return result;
}
