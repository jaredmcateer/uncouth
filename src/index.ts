import { getSingleFileProgram } from './parser'
import { convertAST } from './convert'
import { InputUncouthOptions, getDefaultUncouthOptions, mergeUncouthOptions } from './options'
import { format } from './format'
import path from 'path'
import { readVueSFCOrTsFile, existsFileSync, FileInfo } from './file'
import { setDebugMode } from './debug'
import * as BuiltInPlugins from './plugins/builtIn'

export function convert (content: string, inputOptions: InputUncouthOptions): string {
  const options = mergeUncouthOptions(getDefaultUncouthOptions(inputOptions.typescript), inputOptions)
  const { ast, program } = getSingleFileProgram(content, options)

  return format(convertAST(ast, options, program), options)
}

export function convertFile (filePath: string, root: string, config: string): { file: FileInfo, result: string } {
  root = (typeof root === 'string')
    ? (
      path.isAbsolute(root) ? root : path.resolve(process.cwd(), root)
    )
    : process.cwd()
  config = (typeof config === 'string') ? config : '.uncouth.js'
  if (config.endsWith('.ts')) {
    require('ts-node/register')
  }
  const inputOptions: InputUncouthOptions = existsFileSync(path.resolve(root, config))
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    ? require(path.resolve(root, config)) as InputUncouthOptions
    : {}
  const options = mergeUncouthOptions(getDefaultUncouthOptions(inputOptions.typescript), inputOptions)
  options.root = root

  if (options.debug) {
    setDebugMode(true)
  }

  const file = readVueSFCOrTsFile(filePath, options)
  return {
    file,
    result: convert(file.content, options)
  }
}

export * from './plugins/types'
export { BuiltInPlugins }
export * from './utils'
export { getDefaultUncouthOptions, UncouthOptions } from './options'
