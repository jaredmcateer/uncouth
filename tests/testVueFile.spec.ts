import { convertFile } from "../src";
import path from "path";
import { FileKind } from "../src/file";

describe("testVueFile", () => {
  const filePath = "fixture/Input.vue";

  it("compatible", () => {
    const { file, result } = convertFile(filePath, __dirname, "config/.compatible.uncouth.js");
    expect(file.fsPath.includes(path.basename(filePath))).toBeTruthy();
    expect(path.isAbsolute(file.fsPath)).toBeTruthy();
    expect(file.kind).toBe(FileKind.VUE);
    expect(file).toHaveProperty("start");
    expect(file).toHaveProperty("end");
    expect(result).not.toEqual(/<script>/g);
    expect(result).toMatchSnapshot();
  });

  it("no compatible", () => {
    const { file, result } = convertFile(filePath, __dirname, "config/.nocompatible.uncouth.js");
    expect(file.fsPath.includes(path.basename(filePath))).toBeTruthy();
    expect(path.isAbsolute(file.fsPath)).toBeTruthy();
    expect(file.kind).toBe(FileKind.VUE);
    expect(file).toHaveProperty("start");
    expect(file).toHaveProperty("end");
    expect(result).not.toEqual(/<script>/g);
    expect(result).toMatchSnapshot();
  });
});
