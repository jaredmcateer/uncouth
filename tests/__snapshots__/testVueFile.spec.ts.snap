// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testVueFile compatible 1`] = `
"import MyComponent from \\"my-component.vue\\";
import {
  computed,
  PropType,
  reactive,
  toRefs,
  ref,
  provide,
  inject,
  onBeforeUnmount,
  onUnmounted,
  defineComponent,
} from \\"@vue/composition-api\\";
import { useStore } from \\"vuex\\";
const symbol = Symbol(\\"baz\\");
/**
 * My basic tag
 */
export default defineComponent({
  model: {
    prop: \\"checked\\",
    event: \\"change\\",
  },
  props: {
    bar: null,
    qaq: null,
    cac: null,
    checked: { type: Boolean },
    /**
     * My foo
     */
    foo: { type: Boolean, default: false },
    bar: { type: Number as PropType<number>, default: 1 },
    foobar: { type: Object as PropType<CustomType> },
  },
  name: \\"oao\\",
  setup(props, context) {
    const store = useStore();
    /**
     * My count
     */
    const count = computed(() => {
      return /* TODO: Check this convert result, it can work well in 80% of cases.*/ context
        .root.$store.state.count;
    });
    const { a } = (() => {
      const a = \\"pa\\";
      return toRefs(
        reactive({
          a: a,
        })
      );
    })();
    const anotherComponent = ref<HTMLElement>(null);
    provide(\\"foa\\", \\"foo\\");
    provide(\\"bar\\", \\"bar\\");
    const foai = inject<string>(\\"foai\\");
    const bari = inject<string>(\\"bar\\");
    const optional = inject<string>(\\"optional\\", \\"default\\");
    const bazi = inject<string>(symbol);
    /* TODO: check function dispatch call signature is correct*/ const actA =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"actA\\", arg);
      };
    // hmm all actions are promises.
    const actB = async (str: string): Promise<number> => {
      return await store.dispatch(\\"actB\\", str);
    };
    const actC = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(\\"actC\\", val);
    };
    /* TODO: check function dispatch call signature is correct*/ const actD =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"namespace/actD\\", arg);
      };
    const actE = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(\\"namespace/actE\\", val);
    };
    /* TODO: check function dispatch call signature is correct*/ const actionEff =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(actF, arg);
      };
    const actuhGee = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(actG, val);
    };
    const getA = computed(() => {
      return store.getters.getA;
    });
    const getB = computed<number>(() => {
      return store.getters.getB;
    });
    const getC = computed<(str: string) => number>(() => {
      return store.getters.getC;
    });
    const getTheC = computed(() => {
      return store.getters[\\"namespace/getC\\"];
    });
    const getTheD = computed<number>(() => {
      return store.getters[\\"namespace/getD\\"];
    });
    const getTheE = computed(() => {
      return store.getters.cGetE;
    });
    const getTheG = computed<number>(() => {
      return store.getters.cGetG;
    });
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myDiv =
      ref<HTMLDivElement>(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const mySpan =
      ref(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myComponent =
      ref<MyComponent>(null);
    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.'); //foo
    onBeforeUnmount(() => {
      context.emit(\\"Tearing down\\");
    });
    onUnmounted(() => {
      console.log(\\"destroyed\\");
    });
    const refAccess = () => {
      const foo = { myDiv: true };
      foo.myDiv = false; // should not transform
      myDiv.value!.focus();
      mySpan.value!.innerText = \\"foo\\";
      myComponent.value!.vm.doSomething();
    };
    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value);
      console.log(actB(\\"foo\\"));
      console.log(getC.value(\\"bar\\"));
    };
    // TODO: Please remove unused return variable
    return {
      count,
      a,
      anotherComponent,
      foai,
      bari,
      optional,
      bazi,
      actA,
      actB,
      actC,
      actD,
      actE,
      actionEff,
      actuhGee,
      getA,
      getB,
      getC,
      getTheC,
      getTheD,
      getTheE,
      getTheG,
      myDiv,
      mySpan,
      myComponent,
      msg,
      onBeforeUnmount,
      onUnmounted,
      refAccess,
      hello,
    };
  },
});
"
`;

exports[`testVueFile no compatible 1`] = `
"import MyComponent from \\"my-component.vue\\";
import {
  defineComponent,
  computed,
  PropType,
  reactive,
  toRefs,
  ref,
  provide,
  inject,
  onBeforeUnmount,
  onUnmounted,
} from \\"vue\\";
import { useStore } from \\"vuex\\";
const symbol = Symbol(\\"baz\\");
/**
 * My basic tag
 */
export default defineComponent({
  model: {
    prop: \\"checked\\",
    event: \\"change\\",
  },
  props: {
    bar: null,
    qaq: null,
    cac: null,
    checked: { type: Boolean },
    /**
     * My foo
     */
    foo: { type: Boolean, default: false },
    bar: { type: Number as PropType<number>, default: 1 },
    foobar: { type: Object as PropType<CustomType> },
  },
  name: \\"oao\\",
  setup(props, context) {
    const store = useStore();
    /**
     * My count
     */
    const count = computed(() => {
      return /* TODO: Check this convert result, it can work well in 80% of cases.*/ context
        .root.$store.state.count;
    });
    const { a } = (() => {
      const a = \\"pa\\";
      return toRefs(
        reactive({
          a: a,
        })
      );
    })();
    const anotherComponent = ref<HTMLElement>(null);
    provide(\\"foa\\", \\"foo\\");
    provide(\\"bar\\", \\"bar\\");
    const foai = inject<string>(\\"foai\\");
    const bari = inject<string>(\\"bar\\");
    const optional = inject<string>(\\"optional\\", \\"default\\");
    const bazi = inject<string>(symbol);
    /* TODO: check function dispatch call signature is correct*/ const actA =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"actA\\", arg);
      };
    // hmm all actions are promises.
    const actB = async (str: string): Promise<number> => {
      return await store.dispatch(\\"actB\\", str);
    };
    const actC = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(\\"actC\\", val);
    };
    /* TODO: check function dispatch call signature is correct*/ const actD =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"namespace/actD\\", arg);
      };
    const actE = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(\\"namespace/actE\\", val);
    };
    /* TODO: check function dispatch call signature is correct*/ const actionEff =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(actF, arg);
      };
    const actuhGee = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(actG, val);
    };
    const getA = computed(() => {
      return store.getters.getA;
    });
    const getB = computed<number>(() => {
      return store.getters.getB;
    });
    const getC = computed<(str: string) => number>(() => {
      return store.getters.getC;
    });
    const getTheC = computed(() => {
      return store.getters[\\"namespace/getC\\"];
    });
    const getTheD = computed<number>(() => {
      return store.getters[\\"namespace/getD\\"];
    });
    const getTheE = computed(() => {
      return store.getters.cGetE;
    });
    const getTheG = computed<number>(() => {
      return store.getters.cGetG;
    });
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myDiv =
      ref<HTMLDivElement>(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const mySpan =
      ref(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myComponent =
      ref<MyComponent>(null);
    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.'); //foo
    onBeforeUnmount(() => {
      context.emit(\\"Tearing down\\");
    });
    onUnmounted(() => {
      console.log(\\"destroyed\\");
    });
    const refAccess = () => {
      const foo = { myDiv: true };
      foo.myDiv = false; // should not transform
      myDiv.value!.focus();
      mySpan.value!.innerText = \\"foo\\";
      myComponent.value!.vm.doSomething();
    };
    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value);
      console.log(actB(\\"foo\\"));
      console.log(getC.value(\\"bar\\"));
    };
    // TODO: Please remove unused return variable
    return {
      count,
      a,
      anotherComponent,
      foai,
      bari,
      optional,
      bazi,
      actA,
      actB,
      actC,
      actD,
      actE,
      actionEff,
      actuhGee,
      getA,
      getB,
      getC,
      getTheC,
      getTheD,
      getTheE,
      getTheG,
      myDiv,
      mySpan,
      myComponent,
      msg,
      onBeforeUnmount,
      onUnmounted,
      refAccess,
      hello,
    };
  },
});
"
`;
